File: DesignPattern/structural_patterns/adapter_pattern.py
# Existing class with an incompatible interface
class LegacyComponent:
    def legacy_operation(self):
        print("LegacyComponent is performing its legacy operation.")

# New interface that the client code expects
class NewInterface:
    def new_operation(self):
        pass

# Adapter that adapts the existing class to the new interface
class Adapter(NewInterface):
    def __init__(self, legacy_component):
        self.legacy_component = legacy_component

    def new_operation(self):
        self.legacy_component.legacy_operation()

def main():
    legacy_component = LegacyComponent()
    adapter = Adapter(legacy_component)

    # The client code can work with the new interface
    adapter.new_operation()

if __name__ == "__main__":
    main()



'''
- The Adapter Pattern is a structural design pattern that allows objects with incompatible interfaces to work together. It involves creating 
a wrapper (the adapter) that converts one interface into another. In Python, you can implement the Adapter Pattern easily due to the language's flexibility. 

- In this Python example: LegacyComponent is an existing class with an incompatible interface, and it has a method legacy_operation.

- NewInterface is the new interface that the client code expects, and it has a method new_operation.

- Adapter is the adapter class that inherits from the NewInterface. It contains an instance of LegacyComponent and adapts it to work with the NewInterface.

- In the main function, we create an instance of LegacyComponent, then create an instance of the Adapter and pass the LegacyComponent instance to it. 
Finally, we call the new_operation method of the adapter, which in turn invokes the legacy_operation of the LegacyComponent.

- The Adapter Pattern in Python allows you to use existing classes with different interfaces in your code without modifying the original classes. 
This can be especially useful when working with libraries or legacy code.

'''

File: DesignPattern/structural_patterns/adapter_pattern.cpp
#include <iostream>

// Existing class with an incompatible interface
class LegacyComponent {
public:
    void legacyOperation() {
        std::cout << "LegacyComponent is performing its legacy operation." << std::endl;
    }
};

// New interface that the client code expects
class NewInterface {
public:
    virtual void newOperation() = 0;
    virtual ~NewInterface() {}
};

// Adapter that adapts the existing class to the new interface
class Adapter : public NewInterface {
private:
    LegacyComponent legacyComponent;

public:
    void newOperation() override {
        legacyComponent.legacyOperation();
    }
};

int main() {
    NewInterface* adapter = new Adapter;

    // The client code can work with the new interface
    adapter->newOperation();

    delete adapter;

    return 0;
}



/*
- The Adapter Pattern is a structural design pattern that allows objects with incompatible interfaces to work together. It involves creating 
a wrapper (the adapter) that converts one interface into another. The adapter pattern can be implemented in C++ to make existing classes work 
with others without modifying their source code.

- In this example: LegacyComponent is an existing class with an incompatible interface, and it has a method legacyOperation.

- NewInterface is the new interface that the client code expects, and it has a method newOperation.

- Adapter is the adapter class that derives from the NewInterface and contains an instance of LegacyComponent. It adapts the LegacyComponent 
to work with the NewInterface.

- In the main function, we create an instance of the Adapter and use it to perform the newOperation. The adapter, in turn, invokes the 
legacyOperation of the LegacyComponent.

- The Adapter Pattern allows you to use existing classes with different interfaces in your code without modifying the original classes. This can be 
especially useful when working with libraries or legacy code.
*/